Doc/ai requirements

Segelzwerg opened the pull request:
This is a first draft of the requirements paper for implementing an AI with the help of [BrainAcademy](https://github.com/TreeKid/BrainAcademy) and [ML-Agents](https://github.com/Unity-Technologies/ml-agents).

What is missing:

- [x] <del> Reward System Description </del> (Will be done entirely in BrainAcademy).

- [x] Update doc to changes after the last group call. This includes changes in the observation e.g. map and the necessity of Unity observation must be the same as the input for the neuronal network to work in production.

- [x] Audio Observation

Please feel free to add, what is missing or make PR to this branch if you want to add documentation on it.

Commit:
ADD: first part of ai requirements

Commit:
ADD: heading

Commit:
ADD: about communication requirements

Commit:
ADD: observation requirements

Commit:
ADD: backwards communication

Commit:
ADD: decisions

Commit:
ADD: icon support

Commit:
ADD: export neuronal network

Commit:
UPDATE: passage about game code access

Commit:
UPDATE: reward system

TreeKid commented on doc/AI-requirements.adoc:
```diff
@@ -0,0 +1,170 @@
+time no matter if this choice is valid or possible. So the _game code_ needs to validate the decision. So the neuronal
+network can learn during training, that this decision has no impact and will learn not use it.
+
+There is another approach which catches invalid decisions before hand, but this would only increase the data send from
+and forth between all the parts, so this might not be an option, but could be reconsider during training.
```
Which other approach?

Segelzwerg replied:
I'll could remember, but you were talking something about table before the neuronal network.

TreeKid commented on doc/AI-requirements.adoc:
```diff
@@ -0,0 +1,170 @@
+NPC it will have access to other component.
+
+In addition to the decisions needs to have same dimension every time, the observation must have the same dimension as well.
+This could be tricky for example in mass battles where multiple allies and enemies are around. So in order to get this to
+work we need to cut the observation size to `n`-nearest persons and if are less around fill this vectors with zeros.
```
One simple option is to cut the observation to n nearest persons.
There are also also other more complicated measures.

Segelzwerg replied:
Yes, agreed, this is a part I forgot to revisit after our discussion yesterday.

TreeKid commented on doc/AI-requirements.adoc:
```diff
@@ -0,0 +1,170 @@
+This could be tricky for example in mass battles where multiple allies and enemies are around. So in order to get this to
+work we need to cut the observation size to `n`-nearest persons and if are less around fill this vectors with zeros.
+
+To further reduce the size of the network and there for speed up the training process, the agent will have no memory.
```
Are you talking about the agent from the BrainAcademy or the agent form ML_Agents?

If you are talking about the BrainAcademy agent:
This is only one option and no final decision. 
I definitely want to try agents with memory as well.

Segelzwerg replied:
Yes, I'm talking about the BrainAcademy's Agent. I think it would make no sense to implement a memory in Unity.

All right, I will add this, but my understanding was we don't want to try it. Maybe I missed it.

TreeKid commented on doc/AI-requirements.adoc:
```diff
@@ -0,0 +1,170 @@
+This is mostly relevant for training. It will setup an environment for the training process. It will be controlled from
+`BrainAcademy` and should have no purpose during production.
+
+=== Agent
+
```
Its hard to differentiate between the agent of ML-Agent and the agent of BrainAcademy.
Maybe you could use a more precise naming.

Segelzwerg replied:
Indeed, I will come up with something.

TreeKid reviewed (commented):
Overall I agree with the general concepts.

The list of observations and actions should be specified in detail.
Wen can do this together.

Also the text could be linguistically revised.

Commit:
UPDATE: rework part about data cropping

Commit:
UPDATE: passage about memory

Commit:
UPDATE: reorder definitions alphabetically

Commit:
FIX: cursive

Commit:
FIX: code bracket

Commit:
UPDATE: use explicit agent definition

Segelzwerg commented:
> Overall I agree with the general concepts.

:+1: 

> The list of observations and actions should be specified in detail.
> Wen can do this together.

Yes good idea, maybe in the group call next weekend we should discuss it as well.


> Also the text could be linguistically revised.

That harsh to hear from you :wink: 

I addressed your points, you mentioned.

Commit:
ADD: comment about components

Commit:
ADD: Training paragraph

Commit:
ADD: imitation learning

Commit:
ADD: description of map

Commit:
ADD: important note for attributes

Commit:
UPDATE: environment information

Commit:
ADD: audio observation

iTitus commented on doc/AI-requirements.adoc:
```diff
@@ -0,0 +1,208 @@
+
+====
+
+=== BrainAcademy to Unity ML-Agents and than _game code_
```
```suggestion
=== BrainAcademy to Unity ML-Agents and then _game code_
```

iTitus commented on doc/AI-requirements.adoc:
```diff
@@ -0,0 +1,208 @@
+- Level
+====
+
+Another huge impact of the training will be the information about it's environment. It needs to know, where it's allies
```
```suggestion
Another huge impact of the training will be the information about its environment. It needs to know, where its allies
```

Commit:
FIX: typos

Co-Authored-By: iTitus <iTitus@users.noreply.github.com>

TreeKid committed:
ADD: feature engineering
